### PR TITLE
Potential fix for freeze on exit when Proxy is active.

### LINKED ISSUE #817: Daemon threads in Java preventing Python exit
In ``JPype1<1.0`` user threads created in Java were being terminated early & abruptly if the main thread completed. Since ``JPype1 >= 1.0`` user threads are now honoured.

This seems like a very welcome improvement :+1:, but I didn't see this in the [changelog](https://jpype.readthedocs.io/en/latest/CHANGELOG.html). Was this intentional, and if so, should it be documented?

Code to emulate (excuse my poor Java):

```
package org.hello_world;


public class hello_world {
    public static void thread_hello() throws Exception {
        System.out.println("Setting up a thread");

        Thread t = new Thread(() -> {
            int count = 0;
            while (count < 10) {
                count++;
                try {
                    Thread.currentThread().sleep(1000);
                } catch (InterruptedException e) {
                }
            }
            System.out.println("Thread finished");
        });
        t.setDaemon(false);
        t.start();

        Thread.sleep(750);
        System.out.println("Done in the main thread");
    }
}
```

Which, after compiling with:

```
javac ./org/hello_world/hello_world.java
jar -c ./org/hello_world/hello_world.java ./org/hello_world/hello_world.class > hello-world.jar
```

And running with:

```
import pathlib

import jpype as jp


HERE = pathlib.Path(__file__).parent.absolute()


def test():
    jp.startJVM(f'-Djava.class.path={HERE}/hello-world.jar', convertStrings=False)
    cls = jp.JClass("org.hello_world.hello_world")
    cls.thread_hello()


if __name__ == '__main__':
    test()
```

Can be compared with both JPype 0.7.5 and JPype 1.0.1:

```
$ python -m pip list | grep -i jpype
JPype1                0.7.5               

$ time python run.py 
<java class 'org.hello_world.hello_world'>
Setting up a thread
Done in the main thread

real	0m1.090s
user	0m0.395s
sys	0m0.045s

```

```
$ python -m pip list | grep -i jpype
JPype1                1.0.1

$ time python run.py 
<java class 'org.hello_world.hello_world'>
Setting up a thread
Done in the main thread
Thread finished

real	0m10.428s
user	0m0.468s
sys	0m0.076s

```

Note: I found out about this through some convoluted static declaration in Java which was only hit in a few small cases in pytest tests. The symptom was that if I ran one of these special python tests (with the ``-k`` flag to pytest for instance), I was able to see that pytest ran everything as expected, even reporting the final results, but never actually terminated (so it just sat there without returning the prompt after running the tests). My first instinct was that this was another signal handler issue (e.g. #813) but couldn't find anything relevant in PyTest. As you can see, it turned out to be a complete red-herring, which I might not have fallen for with an relevant entry in the CHANGELOG. Hence the issue :wink:.

### PR DESCRIPTION
**Pull request by Thrameos**
Another attempt at fixing the shutdown sequence.  This defers deleting resources until the JVM is actually shutdown so as survive the race condition that other Java atexit action was still working on a Python thread.  The reference queue is ended but that will die more gracefully.  This still leaves the problem that a daemon thread that was called through Java could still be alive and now has no stack to return to, but that is really a user problem as they should have cleaned up the threads properly.

Fixes #817

### CODE REVIEW

File: native/java/org/jpype/JPypeContext.java [hunk 1]
```diff
@@ -202,15 +202,15 @@ private void shutdown()
...
-      while (this.proxyCount.get() > 0)
-      {
-        try
-        {
-          Thread.sleep(10);
-        } catch (InterruptedException ex)
-        {
-        }
-      }
+//      while (this.proxyCount.get() > 0)
```
Comment (marscher) [inline native/java/org/jpype/JPypeContext.java:205]: Do you have any intentions of reincarnating this code again, or why not just simply delete it?

Comment (Thrameos) [reply native/java/org/jpype/JPypeContext.java:205]: I recommend deleting it, though I am still not sure there isn't a hole in the process.   The shutdown process is always a nightmare if there are still threads in proxy.  I don't think there is a clean way to do it as the two systems don't have great capabilities.   Java doesn't have a way to break a thread with an exception unless it is waiting on IO.   And there is no way to find out which thread belong to Python so we can stop them.

Comment (marscher) [reply native/java/org/jpype/JPypeContext.java:205]: The lack of properly killing Java threads in general is a very unsatisfying situation indeed. So it would not help to hook into thread creation (if that would be possible) and maintain a list of threads created within a proxy (to be able to shut them down somehow).

**Comment by marscher** [comment]
This one needs a rebase on master.

**Comment by Thrameos** [comment]
Okay I will rebase and apply when I have time.  Unfortunately, my development laptop died and the new machine has a driver problem which my support has been unable to fix, so I likely will be down for a week or two.

**Comment by marscher** [comment]
On 8/12/20 3:08 PM, Karl Nelson wrote:
> Okay I will rebase and apply when I have time.  Unfortunately, my development laptop died and the new machine has a driver problem which my support has been unable to fix, so I likely will be down for a week or two.
> 

What a pity with your machine - I hope they sort this problem out quickly.

**Comment by pelson** [comment]
Thanks for this @Thrameos.
I just tested this will all of the cases provided in #817 and confirm that it now behaves as expected. Unfortunately, for the original motivating case, I still get a freeze on termination. I'll try to isolate the case and report it in #817 (or a new issue if that one gets closed). In the meantime I'd say that this was an excellent improvement :+1:

**Comment by pelson** [comment]
I took a good while to track down the issue I'm seeing once and for-all, and it turns out that the problem is in the Java library that I am using, not in JPype (once the change is merged). FWIW my case was related to a class variable starting a thread-pool which created non-daemon threads (as part of a testing suite) - I haven't yet got to the bottom of why this library hasn't had similar "JVM never terminates" type reports, most likely I've missed something with the API that they provide.

So in summary, I endorse this change - it exhibits correct behaviour for all cases that I've tested, and am very happy to see this one addressed! :+1: :tada:
